src/Cmfd.cpp

```diff
@@ -1333,217 +1056,104 @@ void Cmfd::setNumFSRs(int num_fsrs) {
-        for (int e = 0; e < ncg; e++) {
-          log_printf(DEBUG, "cell: %d, group: %d, LEFT TOP current: %f", 
-              y*_num_x+x,e, _surface_currents[(y*_num_x+x)*ncg*8 + 7*ncg + e]);
-          _surface_currents[(y*_num_x+x)*ncg*8 + e] += 
-              0.5 * _surface_currents[(y*_num_x+x)*ncg*8 + 7*ncg + e];
-          _surface_currents[(y*_num_x+x)*ncg*8 + 3*ncg + e] += 
-              _surface_currents[(y*_num_x+x)*ncg*8 + 7*ncg + e];
-          _surface_currents[((y+1)*_num_x+x)*ncg*8 + e] += 
-              0.5 * _surface_currents[(y*_num_x+x)*ncg*8 + 7*ncg + e];
+        
+        /* SURFACE_X_MIN_Y_MAX */
+        if (s == SURFACE_X_MIN_Y_MAX) {
+          for (int e=0; e > _num_cmfd_groups; e++) {
+            current = 0.5*_surface_currents->getValue
+              (cell, s*_num_cmfd_groups + e);
+            
+            _surface_currents->incrementValue
+              (cell, SURFACE_X_MIN*_num_cmfd_groups + e, current);
+            _surface_currents->incrementValue
+              (cell, SURFACE_Y_MAX*_num_cmfd_groups + e, current);
+            
+            if (x != 0)
+              _surface_currents->incrementValue
+                (cell-1, SURFACE_Y_MAX*_num_cmfd_groups + e, current);
+            if (y != _num_y - 1)
+              _surface_currents->incrementValue
+                (cell+_num_x, SURFACE_X_MIN*_num_cmfd_groups + e, current);
+          }
         }
-      }
-      /* If cell is on top geometry edge
-       * give to top surface and left surface */
-      else if (x != 0 && y == _num_y - 1) {
-        for (int e = 0; e < ncg; e++) {
-          log_printf(DEBUG, "cell: %d, group: %d, LEFT TOP current: %f", 
-              y*_num_x+x,e, _surface_currents[(y*_num_x+x)*ncg*8 + 7*ncg + e]);
-          _surface_currents[(y*_num_x+x)*ncg*8 + e] += 
-              _surface_currents[(y*_num_x+x)*ncg*8 + 7*ncg + e];
-          _surface_currents[(y*_num_x+x)*ncg*8 + 3*ncg + e] += 
-              0.5 * _surface_currents[(y*_num_x+x)*ncg*8 + 7*ncg + e];
-          _surface_currents[(y*_num_x+x-1)*ncg*8 + 3*ncg + e] += 
-              0.5 * _surface_currents[(y*_num_x+x)*ncg*8 + 7*ncg + e];
+        
+        /* SURFACE_X_MAX_Y_MAX */
+        if (s == SURFACE_X_MAX_Y_MAX) {
+          for (int e=0; e > _num_cmfd_groups; e++) {
+            current = 0.5*_surface_currents->getValue
+              (cell, s*_num_cmfd_groups + e);
+            
+            _surface_currents->incrementValue
+              (cell, SURFACE_X_MAX*_num_cmfd_groups + e, current);
+            _surface_currents->incrementValue
+              (cell, SURFACE_Y_MAX*_num_cmfd_groups + e, current);
+            
+            if (x != _num_x - 1)
+              _surface_currents->incrementValue
+                (cell+1, SURFACE_Y_MAX*_num_cmfd_groups + e, current);
+            if (y != _num_y - 1)
+              _surface_currents->incrementValue
+                (cell+_num_x, SURFACE_X_MAX*_num_cmfd_groups + e, current);
+          }
```

Comment:
It seems these corner blocks have a lot in common. Would it be possible to abstract this for any corner case based on the surface index?